packages/upload-api/src/ucan/conclude.js

```diff
@@ -0,0 +1,151 @@
+import { equals } from 'uint8arrays/equals'
+
+import { ReferencedInvocationNotFound } from './lib.js'
+
+/**
+ * @param {API.ConcludeServiceContext} context
+ * @returns {API.ServiceMethod<API.UCANConclude, API.UCANConcludeSuccess, API.UCANConcludeFailure>}
+ */
+export const ucanConcludeProvider = ({
+  id,
+  receiptsStorage,
+  tasksStorage,
+  tasksScheduler,
+}) =>
+  provide(conclude, async ({ invocation }) => {
+    const receipt = getConcludeReceipt(invocation)
+
+    // Verify invocation exists failing with ReceiptInvocationNotFound
+    const ranInvocation = receipt.ran
+    const httpPutTaskGetRes = await tasksStorage.get(ranInvocation.link())
+    if (httpPutTaskGetRes.error) {
+      if (httpPutTaskGetRes.error.name === 'RecordNotFound') {
+        return {
+          error: new ReferencedInvocationNotFound(ranInvocation.link()),
+        }
+      }
+      return httpPutTaskGetRes
+    }
+
+    // Store receipt
+    const receiptPutRes = await receiptsStorage.put(receipt)
+    if (receiptPutRes.error) {
+      return {
+        error: receiptPutRes.error,
+      }
+    }
+
+    // THIS IS A TEMPORARY HACK
+    // Schedule `blob/accept` if there is a `http/put` capabilities
+    // inside the invocation that this receipt comes from
+    const scheduleRes = await Promise.all(
+      httpPutTaskGetRes.ok.capabilities
+        // Go through invocation tasks and get all `http/put`
+        .filter((cap) => cap.can === HTTP.put.can)
+        // @ts-expect-error body exists in `http/put` but unknown type here
+        .map(async (/** @type {API.HTTPPut} */ cap) => {
+          // Get triggering task (blob/allocate) by checking blocking task from `url`
+          /** @type {API.UnknownLink} */
+          // @ts-expect-error ts does not know how to get this
+          const [, blobAllocateTaskCid] = cap.nb.url['ucan/await']
+          const blobAllocateTaskGet = await tasksStorage.get(
+            blobAllocateTaskCid
+          )
+          if (blobAllocateTaskGet.error) {
+            return blobAllocateTaskGet
+          }
+
+          /** @type {API.BlobAllocate} */
+          // @ts-expect-error ts does not know how to get this
+          const allocateCapability = blobAllocateTaskGet.ok.capabilities.find(
+            // @ts-expect-error ts does not know how to get this
+            (/** @type {API.BlobAllocate} */ allocateCap) =>
+              equals(allocateCap.nb.blob.digest, cap.nb.body.digest) &&
+              allocateCap.can === W3sBlob.allocate.can
+          )
+
+          const blobAccept = await W3sBlob.accept
+            .invoke({
+              issuer: id,
+              audience: id,
+              with: id.toDIDKey(),
+              nb: {
+                blob: cap.nb.body,
+                space: allocateCapability.nb.space,
+                _put: {
+                  'ucan/await': ['.out.ok', ranInvocation.link()],
+                },
+              },
+              expiration: Infinity,
```

Comment:
I suspect expiry is set to `Infinity` so that CID will come out the same or am I wrong to assume this ? If that is the case indeed can you please add a comment because it's easy to miss and someone may change this without realizing problem it could cause otherwise.